### PR TITLE
Implementing caching in Node.js using redis

### PR DESCRIPTION
**Pull request by mwangiKibui**
### Article Checklist
**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process, quality, and style-once published, you can then submit any number of articles.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by mwangiKibui** [comment]
This fixes #1286

**Comment by adrianmurage** [comment]
@hectorkambow 
kindly run a plagiarism check

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/implementing-caching-in-nodejs-using-redis/index.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+        return res.send({
+            success:false,
+            message:err
+        });
+
+    };    
+
+});
+```
```
Comment (adrianmurage) [inline articles/implementing-caching-in-nodejs-using-redis/index.md:103]: This route does not work.
I added a simple `hello world` route to test if the app itself was functional and it worked but the `/recipe/:foodItem` route doesn't work.
```Javascript
//import express
const express = require("express");
//import axios
const axios = require("axios");
//import redis
const redis = require("redis");

//create an instance of an app.
const app = express();

//configure the port
const PORT = 3001;

//start the server
app.listen(PORT, () => {
  console.log(`app started on ${PORT}`);
});

app.get("/", (req, res) => {
  res.send("Hello World!");
});

app.get(`/recipe/:foodItem`, async (req, res, next) => {
  try {
    //destructure the foodItem from params.
    let { foodItem } = req.params;

    //fetch the data.
    const recipe = await axios
      .get(`http://www.recipepuppy.com/api/?q=${fooditem}`)
      .catch(console.log);

    //return a response.
    return res.send({
      success: true,
      message: recipe.data.results,
    });
  } catch (err) {
    return res.send({
      success: false,
      message: err,
    });
  }
});
```

Attached below is my code currently and a screenshot of my postman request for the "hello world route



Attached below is the postman request for the "/recipe/:foodItem" request

image

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@mwangiKibui 
While trying to follow through with your article, I got stuck.
I've left you a detailed description that should help you understand how to address the issue.

I have installed all the requirements you've pointed out and double checked, so I'm sure that isn't the issue.

I advice that you try and follow through your article from a reader's pov and I'm sure you'll find the issue that way.

@ me when you figure it out.

File: articles/implementing-caching-in-nodejs-using-redis/index.md (hunk 1 above)
Comment (mwangiKibui) [reply articles/implementing-caching-in-nodejs-using-redis/index.md]: fixed

**Comment by mwangiKibui** [comment]
@adrianmurage I have fixed the technical error and some other related errors. Feel free to revisit the article.

File: articles/implementing-caching-in-nodejs-using-redis/index.md [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+```javascript
+const axios = require("axios");
+```
+
+- Modify the `fetchFoodItem` function as follows:
+
+```javascript
+
+const fetchFoodItem = (req,res,next) => {
```
Comment (adrianmurage) [inline articles/implementing-caching-in-nodejs-using-redis/index.md:59]: You ommited the `async` keyword
```suggestion
const fetchFoodItem = async (req,res,next) => {
```

File: articles/implementing-caching-in-nodejs-using-redis/index.md [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+To fetch data from the API we:
+
+- Import `axios` module:
+
+```javascript
+const axios = require("axios");
+```
+
+- Modify the `fetchFoodItem` function as follows:
```
Comment (adrianmurage) [inline articles/implementing-caching-in-nodejs-using-redis/index.md:55]: It's important to point out that the `fetchFoodItem` function is located in `controllers/recipe.js`

File: articles/implementing-caching-in-nodejs-using-redis/index.md [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+redis-server
+```
+
+- If the server is successfully started, continue to the next step. Else if you encounter an error, kindly reference from the [official docs](https://redis.io/topics/quickstart).
+
+### Importing and configuring redis.
+
+Follow the below steps:
+
+- Import `redis` module.
```
Comment (adrianmurage) [inline articles/implementing-caching-in-nodejs-using-redis/index.md:139]: It's important to specify to which file I'm importing Redis.
At the moment, the best I could do is make an educated guess, but that might not be the case for a reader trying to work with Redis for the first time.

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@mwangiKibui 
Your change unblocked me, thank you for that.

But I've faced two other blockers:
1. The `fetchFoodItem` function was missing the `async` keyword.
2. It's not clear where the Redis setup goes in our app and so I got blocked from going forward at that point.

Kindly clarify, for every code change, which files the changes go into.
@ me when you've done that.

Kindly try to follow along with the article from a reader's pov assuming you know nothing. Rely on the article to guide you through the entire process at least once before submitting the changes. This ensures you catch the issues I'm bringing up beforehand, reducing the back and forth.

**Comment by mwangiKibui** [comment]
@adrianmurage I have done the technical fixes.

**Review by adrianmurage — APPROVED** [review]

**Comment by adrianmurage** [comment]
@mwangiKibui 
On the next one consider following through from the point of view of a reader to make sure you catch those small blockers I experienced.
Apart from that, you've done a great job with this, looking forward to reviewing more articles!

**Comment by adrianmurage** [comment]
@hectorkambow 
This article is ready for final review

**Pull request merged**